test/blackbox-tests/test-cases/rule/dependency-external.t/run.t

```diff
@@ -0,0 +1,154 @@
+rules with dependencies outside the build dir are allowed
+
+  $ mkdir -p a/b
+
+  $ cat >a/b/dune-project <<EOF
+  > (lang dune 2.8)
+  > EOF
+
+# Test absolute
+  $ cat >a/b/dune <<EOF
+  > (rule
+  >  (alias test)
+  >  (action (with-stdin-from "$(pwd)/external.txt" (run cat -))))
+  > EOF
+
+  $ cat >external.txt <<EOF
+  > txt1
+  > EOF
+
+  $ dune build --root=a/b @test
+  Entering directory 'a/b'
+  txt1
+
+  $ dune build --root=a/b @test
+  Entering directory 'a/b'
+
+  $ cat >external.txt <<EOF
+  > txt2
+  > EOF
+
+  $ dune build --root=a/b @test
+  Entering directory 'a/b'
+  txt2
+
+# Test copy files absolute
+  $ cat >a/b/dune <<EOF
+  > (rule
+  >  (alias test)
+  >  (action (with-stdin-from "external.txt" (run cat -))))
+  > (copy_files "$(pwd)/external.txt")
+  > EOF
+
+  $ cat >external.txt <<EOF
+  > txt1
+  > EOF
+
+  $ dune build --root=a/b @test
+  Entering directory 'a/b'
+  txt1
+
+  $ dune build --root=a/b @test
+  Entering directory 'a/b'
+
+  $ cat >external.txt <<EOF
+  > txt2
+  > EOF
+
+  $ dune build --root=a/b @test
+  Entering directory 'a/b'
+  txt2
+
+# Test relative 1 level below
+  $ cat >a/b/dune <<EOF
+  > (rule
+  >  (alias test)
+  >  (action (with-stdin-from "../external.txt" (run cat -))))
+  > EOF
+
+  $ cat >a/external.txt <<EOF
+  > txt1
+  > EOF
+
+  $ dune build --root=a/b @test
+  Entering directory 'a/b'
+  txt1
+
+  $ dune build --root=a/b @test
+  Entering directory 'a/b'
+
+  $ cat >a/external.txt <<EOF
+  > txt2
+  > EOF
+
+  $ dune build --root=a/b @test
+  Entering directory 'a/b'
+  txt2
+
+# Test relative 2 level below
+  $ cat >a/b/dune <<EOF
+  > (rule
+  >  (alias test)
+  >  (action (with-stdin-from "../../external.txt" (run cat -))))
+  > EOF
+
+  $ rm -f a/external.txt
+
+  $ cat >external.txt <<EOF
+  > txt1
+  > EOF
+
+  $ dune build --root=a/b @test
+  Entering directory 'a/b'
+  txt1
+
+  $ dune build --root=a/b @test
+  Entering directory 'a/b'
+
+  $ cat >external.txt <<EOF
+  > txt2
+  > EOF
+
+  $ dune build --root=a/b @test
+  Entering directory 'a/b'
+  txt2
+
+# Test copy files 1 level below
+  $ cat >a/b/dune <<EOF
+  > (rule
+  >  (alias test)
+  >  (action (with-stdin-from "external.txt" (run cat -))))
+  > (copy_files ../external.txt)
+  > EOF
+
+  $ cat >a/external.txt <<EOF
+  > txt1
+  > EOF
+
+  $ dune build --root=a/b @test
+  Entering directory 'a/b'
+  File "dune", line 4, characters 12-27:
+  4 | (copy_files ../external.txt)
+                  ^^^^^^^^^^^^^^^
+  Error: path outside the workspace: ../external.txt from .
```

Comment:
I don't understand this error: `../external.txt` should be `a/external.txt` in the context of `a/b/dune`. So it's not outside the workspace. Am I missing something?

Reply:
Oh, got it now.